sdks/java/core/src/main/java/org/apache/beam/sdk/values/BeamRecord.java

```diff
@@ -25,295 +28,340 @@
 import java.util.Date;
 import java.util.GregorianCalendar;
 import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collector;
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import org.apache.beam.sdk.annotations.Experimental;
-import org.apache.beam.sdk.coders.BeamRecordCoder;
 
 /**
  * {@link BeamRecord} is an immutable tuple-like type to represent one element in a
  * {@link PCollection}. The fields are described with a {@link BeamRecordType}.
  *
- * <p>By default, {@link BeamRecordType} only contains the name for each field. It
- * can be extended to support more sophisticated validation by overwriting
- * {@link BeamRecordType#validateValueType(int, Object)}.
- *
- * <p>A Coder {@link BeamRecordCoder} is provided, which wraps the Coder for each data field.
+ * <p>{@link BeamRecordType} contains the names for each field and the coder for the whole
+ * record, {see @link BeamRecordType#getRecordCoder()}.
  */
 @Experimental
 public class BeamRecord implements Serializable {
-  //immutable list of field values.
-  private List<Object> dataValues;
-  private BeamRecordType dataType;
+  private List<Object> values;
+  private BeamRecordType recordType;
 
   /**
-   * Creates a BeamRecord.
-   * @param dataType type of the record
-   * @param rawDataValues values of the record, record's size must match size of
-   *                      the {@code BeamRecordType}, or can be null, if it is null
-   *                      then every field is null.
+   * Creates a {@link BeamRecord} from the list of values and {@link #recordType}.
    */
-  public BeamRecord(BeamRecordType dataType, List<Object> rawDataValues) {
-    if (dataType.getFieldNames().size() != rawDataValues.size()) {
-      throw new IllegalArgumentException(
-          "Field count in BeamRecordType(" + dataType.getFieldNames().size()
-              + ") and rawDataValues(" + rawDataValues.size() + ") must match!");
-    }
+  public static <T> Collector<T, List<Object>, BeamRecord> toRecord(
```

Comment:
can you add a test to test the case when input `BeamRecordType recordType` has more or less fields?